validator/src/journal/publisher.rs

```diff
@@ -1,4 +1,159 @@
-use std::collections::VecDeque;
+use proto::batch::Batch;
+use std::collections::{HashSet, VecDeque};
+use std::slice::Iter;
+use std::sync::Arc;
+use std::sync::mpsc::{channel, Receiver, RecvError, SendError, Sender};
+
+/// This queue keeps track of the batch ids so that components on the edge
+/// can filter out duplicates early. However, there is still an opportunity for
+/// duplicates to make it into this queue, which is intentional to avoid
+/// blocking threads trying to put/get from the queue. Any duplicates
+/// introduced by this must be filtered out later.
+pub fn make_batch_queue() -> (IncomingBatchSender, IncomingBatchReceiver) {
+    let (sender, reciever) = channel();
+    let ids = Arc::new(HashSet::new());
+    (
+        IncomingBatchSender::new(ids.clone(), sender),
+        IncomingBatchReceiver::new(ids, reciever),
+    )
+}
+
+pub struct IncomingBatchReceiver {
+    ids: Arc<HashSet<String>>,
+    receiver: Receiver<Batch>,
+}
+
+impl IncomingBatchReceiver {
+    pub fn new(ids: Arc<HashSet<String>>, receiver: Receiver<Batch>) -> IncomingBatchReceiver {
+        IncomingBatchReceiver { ids, receiver }
+    }
+
+    pub fn get(&mut self) -> Result<Batch, RecvError> {
+        self.receiver.recv().and_then(|batch| {
+            Arc::make_mut(&mut self.ids).remove(batch.get_header_signature());
+            Ok(batch)
+        })
+    }
+}
+
+pub struct IncomingBatchSender {
+    ids: Arc<HashSet<String>>,
+    sender: Sender<Batch>,
+}
+
+impl IncomingBatchSender {
```

Comment:
This would be accessed by the completer. I would expect that something would have an `extern "C"` on the function.

Reply:
Don't worry about that, we'll get it as part of building the FFI layer for these structs